src/Native/System.Native/pal_networking.cpp

```diff
@@ -1028,8 +1033,18 @@ static int32_t GetIPv4PacketInformation(cmsghdr* controlMessage, IPPacketInforma
     }
 
     auto* pktinfo = reinterpret_cast<in_pktinfo*>(CMSG_DATA(controlMessage));
-    ConvertInAddrToByteArray(&packetInfo->Address.Address[0], NUM_BYTES_IN_IPV4_ADDRESS, pktinfo->ipi_addr);
-    packetInfo->InterfaceIndex = static_cast<int32_t>(pktinfo->ipi_ifindex);
+#if HAVE_IN_PKTINFO
+     ConvertInAddrToByteArray(&packetInfo->Address.Address[0], NUM_BYTES_IN_IPV4_ADDRESS, pktinfo->ipi_addr);
+     packetInfo->InterfaceIndex = static_cast<int32_t>(pktinfo->ipi_ifindex);
+#else
+    ConvertInAddrToByteArray(&packetInfo->Address.Address[0], NUM_BYTES_IN_IPV4_ADDRESS, reinterpret_cast<in_addr&>(pktinfo));
```

Comment:
I followed your first suggestion and shrunk the difference to a single statement.